### ISSUE TITLE
Error `type 'Null' is not a subtype of type '_Location' of 'value'` when using tear-off constructor and enabling `Track widget rebuilds` from Devtools

### ISSUE THREAD

**Issue by 30eggis** [flutter/flutter#118368]
Hi, i'm facing this error.

environment is :

IOS/Android
Real Device / Emulators
===============
doctor
```
[✓] Flutter (Channel stable, 3.3.10, on macOS 13.0.1 22A400 darwin-arm
(Rosetta), locale ko-KR)
• Flutter version 3.3.10 on channel stable at /Users/ted/flutter
• Upstream repository https://github.com/flutter/flutter.git
• Framework revision https://github.com/flutter/flutter/commit/135454af32477f815a7525073027a3ff9eff1bfd (4 weeks ago), 2022-12-15 07:36:55 -0800
• Engine revision 3316dd8728
• Dart version 2.18.6
• DevTools version 2.15.0
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
• Android SDK at /Users/ted/Library/Android/sdk
• Platform android-33, build-tools 33.0.0
• ANDROID_HOME = /Users/ted/Library/Android/sdk
• ANDROID_SDK_ROOT = /Users/ted/Library/Android/sdk
• Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
• Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
• All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
• Xcode at /Applications/Xcode.app/Contents/Developer
• Build 14C18
• CocoaPods version 1.11.3

[✓] Chrome - develop for the web
• Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
• Android Studio at /Applications/Android Studio.app/Contents
• Flutter plugin can be installed from:
🔨 https://plugins.jetbrains.com/plugin/9212-flutter
• Dart plugin can be installed from:
🔨 https://plugins.jetbrains.com/plugin/6351-dart
• Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.74.1)
• VS Code at /Applications/Visual Studio Code.app/Contents
• Flutter extension can be installed from:
🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

Scanning for devices is taking a long time...[✓] Connected device (4 available)
• sdk gphone64 arm64 (mobile) • emulator-5554 • android-arm64 •
Android 13 (API 33) (emulator)
• iPhone SE (3rd generation) (mobile) • 111BB95E-72C8-463E-AECA-0D28AB22FB9E • ios •
com.apple.CoreSimulator.SimRuntime.iOS-16-2 (simulator)
• macOS (desktop) • macos • darwin-arm64 •
macOS 13.0.1 22A400 darwin-arm (Rosetta)
• Chrome (web) • chrome • web-javascript •
Google Chrome 108.0.5359.124

[✓] HTTP Host Availability
• All required HTTP hosts are available

• No issues found!
```
And i track inner flutter code, and i found weak null safety.

Error code.
```
======== Exception caught by widgets library =======================================================
The following _TypeError was thrown while rebuilding dirty elements:
type 'Null' is not a subtype of type '_Location' of 'value'

The relevant error-causing widget was:
MediaQuery MediaQuery:file:///Users/ted/project/mobile_app/lib/main.dart:141:36
When the exception was thrown, this was the stack:
#0 List.add (dart:core-patch/growable_array.dart)
https://github.com/flutter/flutter/pull/1 _toLocationId (package:flutter/src/widgets/widget_inspector.dart:3248:14)
https://github.com/flutter/flutter/pull/2 _ElementLocationStatsTracker.add (package:flutter/src/widgets/widget_inspector.dart:2158:20)
https://github.com/flutter/flutter/pull/3 WidgetInspectorService._onRebuildWidget (package:flutter/src/widgets/widget_inspector.dart:2034:19)
https://github.com/flutter/flutter/issues/4 Element.rebuild. (package:flutter/src/widgets/framework.dart:4585:34)
https://github.com/flutter/flutter/issues/5 Element.rebuild (package:flutter/src/widgets/framework.dart:4595:6)
https://github.com/flutter/flutter/issues/6 BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2667:19)
https://github.com/flutter/flutter/issues/7 WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:882:21)
https://github.com/flutter/flutter/pull/8 RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:378:5)
https://github.com/flutter/flutter/issues/9 SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1175:15)
https://github.com/flutter/flutter/pull/10 SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1104:9)
https://github.com/flutter/flutter/pull/11 SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1015:5)
https://github.com/flutter/flutter/issues/12 _invoke (dart:ui/hooks.dart:148:13)
https://github.com/flutter/flutter/pull/13 PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:318:5)
https://github.com/flutter/flutter/pull/14 _drawFrame (dart:ui/hooks.dart:115:31)
The element being rebuilt at the time was index 120 of 430: HomeScreen-[<'HomeScreen'>]
dirty
dependencies: [MediaQuery]
state: _HomeScreenState#b4d12
```

=========

i think weak null safety to below code :

File name : widget_inspector.dart
search keyword : _toLocationId

"_toLocationId" method called from 2 point.
called on method "add" and on method "additionalNodeProperties"

in "additionalNodeProperties" case is null safety, because 'creationLocation' variable should not "null"

```
  @override
  Map<String, Object?> additionalNodeProperties(DiagnosticsNode node) {
    final Map<String, Object?> result = <String, Object?>{};
    final Object? value = node.value;
    if (_interactive) {
      result['objectId'] = service.toId(node, groupName!);
      result['valueId'] = service.toId(value, groupName!);
    }
    if (summaryTree) {
      result['summaryTree'] = true;
    }
    final _Location? creationLocation = _getCreationLocation(value);
    if (creationLocation != null) {
      result['locationId'] = _toLocationId(creationLocation);
    ...
```
but "add" case is not null safety, 'location' variable is no guard of null

```
  void add(Element element) {
    final Object widget = element.widget;
    if (widget is! _HasCreationLocation) {
      return;
    }
    final _HasCreationLocation creationLocationSource = widget;
    final _Location location = creationLocationSource._location;
    final int id = _toLocationId(location);
    ...
```

also, in '_toLocationId()' method is no protect from 'null'

```
int _toLocationId(_Location location) {
  int? id = _locationToId[location];
  if (id != null) {
    return id;
  }
  id = _locations.length;
  _locations.add(location);
  _locationToId[location] = id;
  return id;
}
```

please check this. Thank you

p/s : sorry to my terrible english

**Comment by 30eggis** [comment]
This error occurs when Build - run. 
No occurs when Build 'Release mode'.

**Comment by 30eggis** [comment]
> Hello @30eggis. Thank you for filing this issue. Can you please provide the following information?
> 
> * A [complete and minimal, reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example). Ideally, a single `main.dart` file that can be run directly.
> 
> Please provide the information in the form of text. See how to make collapsible sections with Markdown [here](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections).

Thank's for your reply, but this issue is no occurs on sample project, 
probably  this issue fundamental cause from my mistake. 

but i think "add method" weak null safety is fact, because i think inspector should't disturb my app runtime on debug mode

p/s. this issue occurs on my company project, so i don't upload my original source code.

thank you

**Comment by exaby73** [comment]
Hello. Unfortunately, we need a reproducible code sample in order to take this issue forward. Please try to isolate the issue from your company project and reproduce it in a new project and share the code sample

**Comment by williamhjcho** [comment]
Hey 👋 

we're also experiencing the same issue.
Looking at the latest frames of the stack trace, the call is done within an assertion, which is why it only happens in debug mode.

I'm currently trying different beta versions, but it looks like it is happening at least on `v3.6.0-0.1.pre`

I'm still trying but it's hard to find a minimally reproducible code because where it happens is within multiple complex widgets (w/ slivers in case that's part of the issue), but the final erroring widget is a simple `Text('Invite')`.

**Comment by williamhjcho** [comment]
Ahhh, this is weird:

Apparently, my problem is caused by a combination of these two conditions:
- Flutter Performance's "Track widget rebuilds" is turned on
- Calling a widget constructor via a tear-off 

Here's the minimally reproducible code for my issue:

```dart
import 'package:flutter/material.dart';

void main() {
  return runApp(const IssueApp());
}

class IssueApp extends StatelessWidget {
  const IssueApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Center(
        // 1. Calling the widget as normal
        // child: Text('asd'),

        // 2. Calling a method and creating a widget with a closure is also OK.
        // child: 'asd'.map((str) => Text(str))!,

        // 3. but when using an extension/method that does a tear-off,
        // then the widget rebuild tracker has some issues
        child: 'asd'.map(Text.new)!,
      ),
    );
  }
}

extension MyPackageNullableExtensions<T> on T? {
  R? map<R>(R Function(T) transform) {
    return this != null ? transform(this!) : null;
  }
}
```


- Currently using intellij
- my current flutter SDK version:
  ```
  Flutter 3.7.0-1.4.pre • channel beta • git@github.com:flutter/flutter.git
  Framework • revision 686fe913dc (4 days ago) • 2023-01-10 16:14:01 -0800
  Engine • revision ac5c83dd5f
  Tools • Dart 2.19.0 (build 2.19.0-444.4.beta) • DevTools 2.20.0
  ```

**Comment by exaby73** [comment]
@williamhjcho Are you using the `--track-widget-creation` CLI flag?

**Comment by williamhjcho** [comment]
@exaby73 I'm not interfacing with the CLI directly, just a toggle inside the IDE/flutter plugin itself.

**Comment by exaby73** [comment]
Could you try using it with the CLI and see if it reproduces?

**Comment by williamhjcho** [comment]
Hmm

I tried running it via CLI and nothing happens

But I also just noticed that when the error happens when running on the IDE, it doesn't specifically stop on the "app thread/execution" but on a separate one.
As in, the errors caught by the widget tracking mechanism aren't thrown/shown on the same debugger/output of the app, but as a separate one, that I can only see while it is stopped (on an exception breakpoint).

But while running it by CLI, it just goes through, as if no errors happened.
Maybe it is an issue with the flutter plugin itself? I'm not sure how to debug it further to give more clarification on the issue.

**Comment by exaby73** [comment]
## Triage report

I can reproduce this issue in IntelliJ. But I tried using the Devtools directly, and I am still able to reproduce it, meaning this is not specific to IntelliJ. You can find the `Track widget build counts` option under the `Perfomance` tab, under `More debugging options` in the devtools. I opened the devtools via the terminal, by pressing `v` after running the app with `flutter run`.

<details>
<summary>Logs</summary>

```console
══╡ EXCEPTION CAUGHT BY WIDGETS LIBRARY ╞═══════════════════════════════════════════════════════════
The following _TypeError was thrown building Directionality(textDirection: ltr):
type 'Null' is not a subtype of type '_Location' of 'value'

The relevant error-causing widget was:
  Directionality
  Directionality:file:///Users/nabeelparkar/dev/nevercode/issue_118368/lib/main.dart:12:12

When the exception was thrown, this was the stack:
#0      List.add (dart:core-patch/growable_array.dart)
#1      _toLocationId (package:flutter/src/widgets/widget_inspector.dart:3519:14)
#2      _ElementLocationStatsTracker.add (package:flutter/src/widgets/widget_inspector.dart:2420:20)
#3      WidgetInspectorService._onRebuildWidget (package:flutter/src/widgets/widget_inspector.dart:2296:19)
#4      Element.rebuild.<anonymous closure> (package:flutter/src/widgets/framework.dart:4695:34)
#5      Element.rebuild (package:flutter/src/widgets/framework.dart:4705:6)
#6      StatelessElement.update (package:flutter/src/widgets/framework.dart:5071:5)
#7      Element.updateChild (package:flutter/src/widgets/framework.dart:3684:15)
#8      SingleChildRenderObjectElement.update (package:flutter/src/widgets/framework.dart:6411:14)
#9      Element.updateChild (package:flutter/src/widgets/framework.dart:3684:15)
#10     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5020:16)
#11     Element.rebuild (package:flutter/src/widgets/framework.dart:4715:7)
#12     ProxyElement.update (package:flutter/src/widgets/framework.dart:5328:5)
#13     Element.updateChild (package:flutter/src/widgets/framework.dart:3684:15)
#14     ComponentElement.performRebuild (package:flutter/src/widgets/framework.dart:5020:16)
#15     Element.rebuild (package:flutter/src/widgets/framework.dart:4715:7)
#16     BuildOwner.buildScope (package:flutter/src/widgets/framework.dart:2767:19)
#17     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:864:21)
#18     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:381:5)
#19     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1289:15)
#20     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1218:9)
#21     SchedulerBinding._handleDrawFrame (package:flutter/src/scheduler/binding.dart:1076:5)
#22     _invoke (dart:ui/hooks.dart:145:13)
#23     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:338:5)
#24     _drawFrame (dart:ui/hooks.dart:112:31)

════════════════════════════════════════════════════════════════════════════════════════════════════
```

</details>

<details>
<summary>Code Sample (Same as OP)</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  return runApp(const IssueApp());
}

class IssueApp extends StatelessWidget {
  const IssueApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Directionality(
      textDirection: TextDirection.ltr,
      child: Center(
        // 1. Calling the widget as normal
        // child: Text('asd'),

        // 2. Calling a method and creating a widget with a closure is also OK.
        // child: 'asd'.map((str) => Text(str))!,

        // 3. but when using an extension/method that does a tear-off,
        // then the widget rebuild tracker has some issues
        child: 'asd'.map(Text.new)!,
      ),
    );
  }
}

extension MyPackageNullableExtensions<T> on T? {
  R? map<R>(R Function(T) transform) {
    // ignore: null_check_on_nullable_type_parameter
    return this != null ? transform(this!) : null;
  }
}
```

</details>

<details><summary><code>flutter doctor -v</code> (Master)</summary>

```
[!] Flutter (Channel master, 3.7.0-21.0.pre.45, on macOS 13.1 22C65 darwin-arm64, locale en-US)
    • Flutter version 3.7.0-21.0.pre.45 on channel master at /Users/nabeelparkar/fvm/versions/master
    ! Warning: `dart` on your path resolves to /opt/homebrew/Cellar/dart/2.18.6/libexec/bin/dart, which is not inside your current Flutter SDK checkout at /Users/nabeelparkar/fvm/versions/master. Consider adding /Users/nabeelparkar/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f22280a0c8 (64 seconds ago), 2023-01-16 22:09:06 -0800
    • Engine revision f790304409
    • Dart version 3.0.0 (build 3.0.0-122.0.dev)
    • DevTools version 2.20.0
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nabeelparkar/Library/Android/sdk/
    • Platform android-33, build-tools 33.0.0
    • ANDROID_SDK_ROOT = /Users/nabeelparkar/Library/Android/sdk/
    • Java binary at: /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • CHROME_EXECUTABLE = /Applications/Brave Browser.app/Contents/MacOS/Brave Browser

[✓] Android Studio (version 2021.3)
    • Android Studio at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/213.7172.25.2113.9123335/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.1)
    • IntelliJ at /Users/nabeelparkar/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 71.3.6
    • Dart plugin version 223.8214.16

[✓] IntelliJ IDEA Ultimate Edition (version 2022.3.1)
    • IntelliJ at /Users/nabeelparkar/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/223.8214.52/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.74.3)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.56.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.1 22C65 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Brave Browser 109.1.47.171

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

</details>
